File: binding/php/search_test.php
<?php
// Copyright 2022 The Ip2Region Authors. All rights reserved.
// Use of this source code is governed by a Apache2.0-style
// license that can be found in the LICENSE file.
//
// @Author Lion <chenxin619315@gmail.com>
// @Date   2022/06/21

require dirname(__FILE__) . '/xdb/Searcher.class.php';

use \ip2region\xdb\Util;
use \ip2region\xdb\{IPv4, IPv6};
use \ip2region\xdb\Searcher;

function printHelp($argv) {
    printf("php %s [command options]\n", $argv[0]);
    printf("options: \n");
    printf(" --db string             ip2region binary xdb file path\n");
    printf(" --cache-policy string   cache policy: file/vectorIndex/content\n");
}

if($argc < 2) {
    printHelp($argv);
    return;
}

$dbFile = "";
$cachePolicy = 'vectorIndex';
array_shift($argv);
foreach ($argv as $r) {
    if (strlen($r) < 5) {
        continue;
    }

    if (strpos($r, '--') != 0) {
        continue;
    }

    $sIdx = strpos($r, "=");
    if ($sIdx < 0) {
        printf("missing = for args pair %s\n", $r);
        return;
    }

    $key = substr($r, 2, $sIdx - 2);
    $val = substr($r, $sIdx + 1);
    if ($key == 'db') {
        $dbFile = $val;
    } else if ($key == 'cache-policy') {
        $cachePolicy = $val;
    } else {
        printf("undefined option `%s`\n", $r);
        return;
    }
}

if (strlen($dbFile) < 1) {
    printHelp($argv);
    return;
}

// printf("debug: dbFile: %s, cachePolicy: %s\n", $dbFile, $cachePolicy);
$handle = fopen($dbFile, 'r');
if ($handle === false) {
    printf("failed to open the xdb file `{$dbFile}`\n");
    return;
}

// verify the xdb file
// @Note: do NOT call it every time you create a searcher since this will slow
// down the search response.
// @see the Util.verify function for details.
$err = Util::verify($handle);
if ($err != null) {
    printf("failed to verify xdb file `%s`: %s\n", $dbFile, $err);
    return;
}

// load header
$header = Util::loadHeader($handle);
if ($header == null) {
    printf("failed to load the header\n");
    return;
}

// get the version number from the xdb header
try {
    $version = Util::versionFromHeader($header);
} catch (Exception $e) {
    printf("failed to detect version from header: {$e->getMessage()}\n");
    return;
}

// create the xdb searcher by the cache-policy
switch ( $cachePolicy ) {
case 'file':
    try {
        $searcher = Searcher::newWithFileOnly($version, $dbFile);
    } catch (Exception $e) {
        printf("failed to create searcher with '%s': %s\n", $dbFile, $e);
        return;
    }
    break;
case 'vectorIndex':
    $vIndex = Util::loadVectorIndex($handle);
    if ($vIndex == null) {
        printf("failed to load vector index from '%s'\n", $dbFile);
        return;
    }

    try {
        $searcher = Searcher::newWithVectorIndex($version, $dbFile, $vIndex);
    } catch (Exception $e) {
        printf("failed to create vector index cached searcher with '%s': %s\n", $dbFile, $e);
        return;
    }
    break;
case 'content':
    $cBuff = Util::loadContent($handle);
    if ($cBuff == null) {
        printf("failed to load xdb content from '%s'\n", $dbFile);
        return;
    }

    try {
        $searcher = Searcher::newWithBuffer($version, $cBuff);
    } catch (Exception $e) {
        printf("failed to create content cached searcher: %s", $e);
        return;
    }
    break;
default:
    printf("undefined cache-policy `%s`\n", $cachePolicy);
    return;
}

printf(<<<EOF
ip2region xdb searcher test program
source xdb file: {$dbFile} ({$version->name}, ${cachePolicy})
type 'quit' to exit\n
EOF);
while ( true ) {
    echo "ip2region>> ";
    $line = trim(fgets(STDIN));
    if (strlen($line) < 2) {
        continue;
    }

    if ($line == 'quit') {
        break;
    }

    $cost = -1;
    try {
        $sTime = Util::now();
        $region = $searcher->search($line);
        $cost = Util::now() - $sTime;
    } catch (Exception $e) {
        printf("search call failed: %s\n", $e->getMessage());
        continue;
    }

    printf(
        "{region: %s, ioCount: %d, took: %.5f ms}\n",
        $region, $searcher->getIOCount(), $cost
    );
}

// close the searcher at last
$searcher->close();
printf("searcher test program exited, thanks for trying\n");

File: binding/python/search_test.py
# Copyright 2022 The Ip2Region Authors. All rights reserved.
# Use of this source code is governed by a Apache2.0-style
# license that can be found in the LICENSE file.

# xdb searcher test on 2025/10/30
# Author Leon<chenxin619315@gmail.com>

import io
import sys
import argparse
import time
import ip2region.util as util
import ip2region.searcher as xdb

def create_searcher(db_path, cache_policy):
    # open the source xdb file
    handle = io.open(db_path, "rb")

    # verify the xdb file
    # @Note: do NOT call it every time you create a searcher since this will slow
    # down the search response.
    # @see the verify function for details.
    util.verify(handle)

    # get the ip version from header
    header = util.load_header(handle)
    version = util.version_from_header(header)
    if version is None:
        handle.close()
        raise Exception("failed to get version from header")

    searcher = None
    if cache_policy == "file":
        searcher = xdb.new_with_file_only(version, db_path)
    elif cache_policy == "vectorIndex":
        v_index = util.load_vector_index(handle)
        searcher = xdb.new_with_vector_index(version, db_path, v_index)
    elif cache_policy == "content":
        c_buffer = util.load_content(handle)
        searcher = xdb.new_with_buffer(version, c_buffer)
    else:
        raise ValueError("invalid cache_policy `{}`".format(cache_policy))

    handle.close()
    return searcher


def run_search_test(db_path: str, cache_policy: str):
    # create the searcher
    searcher = None
    try:
        searcher = create_searcher(args.db, args.cache_policy)
    except Exception as e:
        print("failed to create searcher: {}".format(str(e)))
        return

    # print the searcher for debug
    # print("searcher -> ", searcher)
    print('''ip2region xdb searcher test program
source xdb: {} ({}, {})
type 'quit' to exit'''.format(db_path, searcher.get_ip_version().name, cache_policy))

    # get input ip address and do the search
    while True:
        ip_str = input("ip2region>> ").strip()

        if len(ip_str) < 2:
            continue
        if ip_str == "quit":
            break

        s_time = time.time()
        try:
            ip_bytes = util.parse_ip(ip_str)
        except Exception as e:
            print(f"invalid ip address `{ip_str}`")
            continue
        
        try:
            region = searcher.search(ip_bytes)
        except Exception as e:
            print("failed to search({}): {}".format(util.ip_to_string(ip_bytes), str(e)))
            continue

        took = (time.time() - s_time) * 1000
        print(f"{{region: {region}, ioCount: {searcher.get_io_count()}, took: {took:.3f} ms}}");

    # close the searcher
    searcher.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        add_help=True,
        prog="python search_test.py",
        description="ip2region search test script",
        usage="%(prog)s [command option]"
    )

    # check the args
    parser.add_argument('--db', help='ip2region binary xdb file path')
    parser.add_argument('--cache-policy', help='cache policy: file/vectorIndex/content, default: vectorIndex', default="vectorIndex")
    args = parser.parse_args()
    if args.db is None:
        parser.print_help()
        sys.exit()

    # run the search test
    run_search_test(args.db, args.cache_policy)